### DISCUSSION TITLE
Compile to dll

### DISCUSSION THREAD

**Discussion by Tampa** [xunit/xunit#2764] [Question] [answered]
Project setup uses a custom prebuild and thus cannot use dotnet functions to add xunit to the project files. Instead needs a dll to reference the namespace directly and provide it to the generated project files. Is it possible to build xunit into a dll to provide the entire namespace as reference much like nunit.framework.dll does?

**Comment by bradwilson** [comment, marked as answer]
I don't understand the request.

We ship DLLs in our NuGet packages. We do not ship DLLs in any other way. If the request is for us to ship DLLs that aren't in NuGet packages, we are not planning to support that.

**Reply by Tampa** [reply]
Perhaps I phrased it a bit complicated.
Previously I would simply import the nunit namespace by direct reference to the dll: <Reference name="nunit.framework"/> This would allow the prebuild to add it to the project files for each test project in the solution.
Now I would like to switch to xunit referencing a similar dll so prebuild can produce the project files, but I cannot find a dll that does this. I checked the nuget packages, but I did not see a xunit.dll or something that would provide the namespace. The documentation talks about using dotnet commands to add it directly to projects, but I cannot do this in this setup, has to be some dll.

In most basic form I am looking for the equivalent dll that can provide what nunit.framework.dll provided just for xunit. Is that in one of the nuget packages or can I build such a dll?

**Reply by bradwilson** [reply]
To build unit tests, you would need these DLL references:

- `xunit.assert.dll` from `xunit.assert` NuGet package
- `xunit.core.dll` from `xunit.extensibility.core` NuGet Package
- `xunit.execution.desktop.dll` (for .NET Framework) or `xunit.execution.dotnet.dll` (for .NET Core) from `xunit.extensibility.execution` NuGet packages

You would also need to import via MSBuild:

- `xunit.core.props` and `xunit.core.targets` from `xunit.core` NuGet package

Note: I've never tried this and don't know if it actually works, but that's effectively what happens when you reference the `xunit` NuGet package (plus you get the Roslyn analyzers, but there's no way to get those without NuGet as far as I know, because they're not direct references).

**Reply by Tampa** [reply]
I am not sure what you mean by import via MSBuild, but the dll references seem to work, I was able to use the Xunit namespace and write a basic test and it even compiled into a dedicated dll. So much thanks for that! :)

The Test Explorer in vs also found the test, but I cannot run it directly it seems. Initially the problem is:
Test project Tests does not reference any .NET NuGet adapter. Test discovery or execution might not work for this project.
It's recommended to reference NuGet test adapters in each test project in the solution.

Adding the nuget package for "Microsoft.NET.Test.SDK" then throws Could not find testhost, which is fixed by manually changing the project file. The next issue is the missing TestPlatform dlls, which I can supply, but are required to use an older version, which subsequently visual studio can't seem to work with. The target is dotnet 6 so perhaps it is already all geared to dotnet 7 causing this to fail.

`System.IO.FileNotFoundException: Could not load file or assembly 'Microsoft.TestPlatform.PlatformAbstractions, Version=15.0.0.0`

While the aim is to run them via console it would be very useful to be able to run them in visual studio. I am at a bit of a loss as to why so much nuget stuff is necessary to run some tests. I can manually add the required dll to the prebuild I suppose, but that does nothing if visual studio then fails to run the tests. Can I downgrade the test runner in visual studio somehow or what would be the course of action here you think?

**Reply by bradwilson** [reply]
> I am not sure what you mean by import via MSBuild

You add `<Import Project="path/to/filename" />` to the top of your MSBuild file (typically your .csproj file). https://learn.microsoft.com/en-us/visualstudio/msbuild/import-element-msbuild?view=vs-2022

**Reply by bradwilson** [reply]
> The Test Explorer in vs also found the test, but I cannot run it directly it seems.

You'll notice that our `xunit.runner.visualstudio` package doesn't have libraries for you to link against; instead it has a `build` folder with a .props file (which is linked in via `Import` automatically), which copies the DLLs in question into your output folder when building. It's _possible_ that ensuring that you add the `Import` for the `.props` file will make Test Explorer work, but I'm a little bit skeptical about that. It's something I've never tried (and like everything discussed so far, something we don't support).

> I am at a bit of a loss as to why so much nuget stuff is necessary to run some tests.

This was the design of Test Explorer. It's one of the only two supported mechanisms for test runners (the other being a VSIX-based distribution, which was stopped shipping many years ago when they deprecated it in favor of the NuGet-based solution).

To be honest: you're swimming _way upstream_ by avoiding NuGet. Finding things that don't work is going to be the norm, and you trying to patch things together by hand might work and it might not. You're almost surely going to find that nobody on the Microsoft side is going to offer you any support. Even me answering this questions, I'm making clear that I'm giving you a best guess on something that might or might not work, but even if it works, you're on your own. Anything that's broken is on you to fix.

**Reply by Tampa** [reply]
So I figured out that I need some older versions of things. I now installed MSTest.TestAdapter 2.2.3 and Microsoft.NET.Test.Sdk 16.7.0 This seems to fix all the previous errors I was getting and replaced it with: `Microsoft.VisualStudio.TestPlatform.ObjectModel.TestPlatformException: Failed to negotiate protocol` which I assume means the version of xunit is not compatible and likely too new? Tried all the latest versions those were big fail, this seems closer to victory, apart from not really knowing how I am going to teach prebuild to integrate the nuget packages. Thanks for all your help thus far, greatly appreciated! :)

--- EDIT 2 ---

I certainly appreciate all the help. I'm not the one calling the shots with the project itself, just trying to help out building a test setup. While I'm not personally a fan of the whole principle of nuget and similar package management for development, I'm fine to use it for getting local testing to work, but I also have to think about the portability of the code when tests have to be run from console. I apologize if I am being a bother, I never did expect an answer so what you have provided already helped get things further. I'll try to reach out to other members of the project to see if they have an idea on how to get it working.

--- EDIT ---

The csproj files are not static, they are created via prebuild instead, so normally aren't changed in any way. That's why I have to reference everything via the dlls rather than being able to use nuget for it.

I'm still a bit confused as to why visual studio seems to find the test, but then reports it cannot find the test container when the dll exists.
One of the suggested solutions, adding Microsoft.NET.Test.SDK via nuget then shifts to `Could not load file or assembly 'Microsoft.TestPlatform.PlatformAbstractions, Version=15.0.0.0`, which I tried to add, but it just seems to result in version mismatches and like I said, can't use nuget or manual edits to project files.

I tried running the test via `dotnet test Tests.dll` and it reports:

```
Microsoft (R) Test Execution Command Line Tool Version 17.6.3 (x64)
Copyright (c) Microsoft Corporation.  All rights reserved.

Starting test execution, please wait...
A total of 1 test files matched the specified pattern.
No test is available in G:\$GitLab$\#####\Tests\Tests.dll. Make sure that test discoverer & executors are registered and platform & framework version settings are appropriate and try again.

Additionally, path to test adapters can be specified using /TestAdapterPath command. Example  /TestAdapterPath:<pathToCustomAdapters>.
```

If visual studio requires a specific version of the test sdk and testplatform via nuget that's fine, but ultimately a Linux box has to be able to run these tests as well as the whole project gets built and tested by a Jenkins instance at the moment.

For net6 target which versions should I use? (Already tried 15.6.0 and the latest) Or is there something I'm missing?

**Reply by bradwilson** [reply]
> I'm still a bit confused as to why visual studio seems to find the test, but then reports it cannot find the test container when the dll exists.

You very likely have this check box checked, since it is checked by default:

![image](https://github.com/xunit/xunit/assets/16944/8275c12b-fd65-4587-b3bb-648834d9a117)

That runs some Microsoft-owned code which is independent of our VSTest adapter. It sees the reference to xUnit.net and tries to guess which source code looks like a test and which doesn't. That doesn't help it run (or discover after build). This dichotomy between what Visual Studio thinks and what is actually true is why I have consistently told people to uncheck this check box any time they have any discrepancy (like you're experiencing). I'd be super surprised if unchecking that box (and restarting VS) doesn't solve this mystery.

> Additionally, path to test adapters can be specified using /TestAdapterPath command

This might solve your command line problem. You need to point to the folder where `xunit.runner.visualstudio.testadapter.dll` (for .NET Framework test projects) or `xunit.runner.visualstudio.dotnetcore.testadapter.dll` (for .NET Core test projects) lives. This probably isn't working for you right now because you're missing the MSBuild Import statements to the .props file in the `xunit.runner.visualstudio` NuGet package, as that's what causes those DLLs to be copied into your output folder during build (and would then allow the command line to find the adapter implicitly).

> If visual studio requires a specific version of the test sdk and testplatform via nuget that's fine

I have no idea what versions it needs of which things, and which come via NuGet vs. being on disk already. Obviously they have optimized their system for the NuGet-based delivery of test adapters, and don't really document their internal processes (as they likely see that to be unnecessary, and wouldn't want to document something that they want to be free to change without notice).

**Reply by Tampa** [reply]
I did manage to get to a point where it appears everything is in order and things are found, but when attempting to run the test I just get:

`Microsoft.VisualStudio.TestPlatform.ObjectModel.TestPlatformException: Failed to negotiate protocol`

after a while. Looking up what the cause of that might be I find a lot of results that ran into this after upgrading their solutions. This suggests to me that I have a version mismatch somewhere. Unfortunately there isn't more verbose data as to what version is expects or anything like that. I guess I have to try different versions of TestPlatform and hope to find the correct one or try the other packages if that fails. I tried matching everything to netstandard1.1, but TestPlatform shows net451 still even in the latest version so not sure if that's correct. Trial and error I guess.

Thanks for all the help thus far, I guess I'll just have to keep trying to find the right versions of things. :)

**Reply by bradwilson** [reply]
I'm going to mark this thread as the answer. If you come up with something that works and want to document it for anybody else in this situation, feel free to come back and provide the info. Thanks!
